### PR TITLE
HTTP/3 Made alt-used show only on alt-svc redirect

### LINKED ISSUE #56876: Http/3 adds alt-used header even when alt-svc isn't is used
https://github.com/dotnet/runtime/blob/0c200955af77bf17cd05488db96d3cdb3716a16c/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs#L528-L529

This code makes the incorrect assumption that Alt-Svc is required, and therefore Alt-Used should always be added. Alt-Svc isn't required when using `HttpVersion.Version30` and `HttpVersionPolicy.RequestVersionExact` and Alt-Used should not be added in this situation.

```c#
            using var client = new HttpClient(handler);
            client.DefaultRequestVersion = HttpVersion.Version30;
            client.DefaultVersionPolicy = HttpVersionPolicy.RequestVersionExact;
            var response = await client.GetStringAsync(address);
```

Impact: Unknown
Priority: Low

### PR DESCRIPTION
**Pull request by iuliaco**
Fixes #56876
I only did a test for checking not alt-used included, tests which include checking alt-used header are already implemented [here](https://github.com/dotnet/runtime/blob/a5e678dd58c365764d0889e137f5953533dbb451/src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.AltSvc.cs#L125).

### CODE REVIEW

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs [hunk 1]
```diff
@@ -70,16 +70,22 @@ private bool ShuttingDown
...
-            _altUsedEncodedHeader = QPack.QPackEncoder.EncodeLiteralHeaderFieldWithoutNameReferenceToArray(KnownHeaders.AltUsed.Name, altUsedValue);
+
+
+            if (altUsedEncode)
+            {
+                bool altUsedDefaultPort = pool.Kind == HttpConnectionKind.Http && authority.Port == HttpConnectionPool.DefaultHttpPort || pool.Kind == HttpConnectionKind.Https && authority.Port == HttpConnectionPool.DefaultHttpsPort;
+                string altUsedValue = altUsedDefaultPort ? authority.IdnHost : string.Create(CultureInfo.InvariantCulture, $"{authority.IdnHost}:{authority.Port}");
+                _altUsedEncodedHeader = QPack.QPackEncoder.EncodeLiteralHeaderFieldWithoutNameReferenceToArray(KnownHeaders.AltUsed.Name, altUsedValue);
+            }
+
```
Comment (ManickaP) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs:88]:
```suggestion
            if (altUsedEncode)
            {
                bool altUsedDefaultPort = pool.Kind == HttpConnectionKind.Http && authority.Port == HttpConnectionPool.DefaultHttpPort || pool.Kind == HttpConnectionKind.Https && authority.Port == HttpConnectionPool.DefaultHttpsPort;
                string altUsedValue = altUsedDefaultPort ? authority.IdnHost : string.Create(CultureInfo.InvariantCulture, $"{authority.IdnHost}:{authority.Port}");
                _altUsedEncodedHeader = QPack.QPackEncoder.EncodeLiteralHeaderFieldWithoutNameReferenceToArray(KnownHeaders.AltUsed.Name, altUsedValue);
            }
```
White spaces, and please fix the parameter name.

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs [hunk 2]
```diff
@@ -972,6 +972,39 @@ public async Task Alpn_H3_Success()
...
+            using Http3LoopbackServer server = CreateHttp3LoopbackServer();
+
+            Http3LoopbackConnection connection = null;
+            HttpRequestData requestData = null;
+            Task serverTask = Task.Run(async () =>
+            {
+                connection = (Http3LoopbackConnection)await server.EstablishGenericConnectionAsync();
+                requestData = await connection.ReadRequestDataAsync(readBody: false);
+                Assert.NotNull(connection);
+                Assert.Throws<System.Exception>(() => requestData.GetSingleHeaderValue("alt-used"));
```
Comment (CarnaViire) [inline src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs:987]:
```suggestion
                Assert.Equal(0, requestData.GetHeaderValueCount("alt-used"));
```

**Review by CarnaViire — APPROVED** [review]
LGTM modulo nits, thanks!

**Comment by CarnaViire** [comment]
@iuliaco could you please update your branch to the current main (because https://github.com/dotnet/runtime/pull/72854 is merged now), so we could rerun the tests in CI without Http tests being timed out

**Comment by iuliaco** [comment]
> 

Done

**Pull request merged**